src/Helper/Escaper/AbstractHelper.php

```diff
@@ -89,9 +100,12 @@ abstract protected function escape($value);
     /**
      * Set the encoding to use for escape operations
      *
-     * @param  string $encoding
+     * @deprecated Will be removed in 3.0.
```

Comment:
@gsteel @Ocramius 
This is a no-go, because as long as these methods are present and the class is not final, it can be overwritten. Therefore, we must assume that this will be done.
This means when a deprecation is introduced that we also have to explain what the alternative is. The current description is therefore insufficient.
(It would be perfect if we could add a link to the documentation with further explanations and an example.)

Reply:
The idea is to drop the abstract class overall? 🤔

Reply:
@froschdesign Won't the migration guide cover this?

Should the migration guide to 3 be seeded in 2.x so that the docs can be written and referenced for each deprecation?

Reply:
> The idea is to drop the abstract class overall?

I was actually referring to `Helper\AbstractHelper` in the pull description. `Helper\Escaper\AbstractHelper` IMO could be concrete and final so that the user can choose:

```php
// Single Helper for all escaping contexts
$this->escaper()->escapeJs('foo');
$this->escaper()->escapeHtml('foo');

// Existing helpers for convenience, or maybe just get rid of them?
$this->escapeHtml('foo'); // etc...
```

The useful stuff in `Helper\Escaper\AbstractHelper` with regard to recursion could be extracted into a trait for re-use.

Reply:
> Won't the migration guide cover this?

So far this is not the case and if the migration guide covers this, the link should be added here.

I see two options for deprecations like this:

1. add an explanation of what the alternative is / what the user should do instead in the DocBlock
2. create a migration guide like "Migration from Version 2.x to 2.20" and add the link to the DocBlock

The goal must be that the user does not have to search or ask. The past has shown that questions arise in the forum, in the chat and new issue reports if we do not clarify things properly.
Thank you for understanding!

(We must not forget the release notes too, the explanations and / or the link must also be added there.)